### PR TITLE
Update prerelease bugs

### PR DESCRIPTION
**Pull request by johnataylor**
- couple of bugs on Update-Package with prerelease packages in the project
- new "multi" api for PackageManagement whereby the caller can give a set of PackageIdentities to resolve

### CODE REVIEW

**Comment by johnataylor** [comment]
relevant NuGet/Home issues 1545, 1370 and 1466

File: src/NuGet.Core/NuGet.PackageManagement/Resolution/PrunePackageTree.cs [hunk 1]
```diff
@@ -16,6 +16,29 @@ namespace NuGet.PackageManagement
...
+        /// </summary>
+        public static IEnumerable<SourcePackageDependencyInfo> PrunePrereleaseExceptAllowed(IEnumerable<SourcePackageDependencyInfo> packages, IEnumerable<PackageIdentity> installedPackages, bool isUpdateAll)
+        {
+            var allowedPackageIdentity = new HashSet<PackageIdentity>(installedPackages.Where(p => p.HasVersion && p.Version.IsPrerelease), PackageIdentityComparer.Default);
+
+            if (isUpdateAll)
+            {
+                // If this is an Update All scenario then we will allow package that are already prerelease to pick any other prerelease alternatievs
+
+                var allowedPackageId = new HashSet<string>(allowedPackageIdentity.Select(p => p.Id), StringComparer.OrdinalIgnoreCase);
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.PackageManagement/Resolution/PrunePackageTree.cs:29]: This should be able to use a hashset of all installed packages, not just prerelease.

**Comment by emgarten** [comment]
:shipit:

**Comment by johnataylor** [comment]
merged from command line

**Pull request closed without merging**